### PR TITLE
Support SFTP as file location

### PR DESCRIPTION
**Pull request by rajaths010494**
**Please describe the feature you'd like to see**
It is a very common ETL operation to either push data to or load data from FTP sources

**Describe the solution you'd like**
SFTP should be valid location types in `File`

**Are there any alternatives to this feature?**
Implementors otherwise need to use workarounds like `S3toSFTPOperator`

**Acceptance Criteria**

- [x] All checks and tests in the CI should pass
- [x] Unit tests (90% code coverage or more, [once available](https://github.com/astronomer/astro-sdk/issues/191))
- [x] Integration tests (if the feature relates to a new database or external service)
- [x] Example DAG
- [x] Docstrings in [reStructuredText](https://peps.python.org/pep-0287/) for each of  methods, classes, functions and module-level attributes (including Example DAG on how it should be used)
- [x] Exception handling in case of errors
- [x] Logging (are we exposing useful information to the user? e.g. source and destination)
- [x] Improve the documentation (README, Sphinx, and any other relevant)
- [x] How to use Guide for the feature ([example](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/operators/postgres_operator_howto_guide.html))

### CODE REVIEW

File: .github/ci-test-connections.yaml [hunk 1]
```diff
@@ -94,3 +94,10 @@ connections:
     password: $DATABRICKS_TOKEN
     extra:
       http_path: /sql/1.0/warehouses/cf414a2206dfb397
+  - conn_id: sftp_conn
+    conn_type: sftp
+    host: localhost
+    login: foo
+    password: pass
```
Comment (pankajkoti) [inline .github/ci-test-connections.yaml:101]: This should be env vars

File: python-sdk/example_dags/example_load_file.py [hunk 2]
```diff
@@ -253,4 +253,17 @@
     )
     # [END load_file_example_19]
 
+    # [START load_file_example_20]
+    aql.load_file(
+        input_file=File(path="sftp://foo@localhost/upload/", conn_id="sftp_conn", filetype=FileType.CSV),
```
Comment (pankajkoti) [inline python-sdk/example_dags/example_load_file.py:258]: username and hostname could be env vars, here?

File: python-sdk/src/astro/files/locations/sftp/sftp.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            key_file = extra_options.get("key_file")
+            return {"connect_kwargs": {"key_filename": key_file}}
+        elif client.password:
+            return {"connect_kwargs": {"password": client.password}}
+        return {}
+
+    @property
+    def paths(self) -> list[str]:
+        """Resolve SFTP file paths with prefix"""
+        url = urlparse(self.path)
```
Comment (pankajkoti) [inline python-sdk/src/astro/files/locations/sftp/sftp.py:36]: We're parsing this multiple times in each of the properties. Can we have a parsed url as an instance variable and reuse it?

File: .github/ci-test-connections.yaml (hunk 1 above)
Comment (rajaths010494) [reply .github/ci-test-connections.yaml:101]: Need to add in github env and pull it.

Comment (rajaths010494) [reply .github/ci-test-connections.yaml]: modified

File: python-sdk/example_dags/example_load_file.py (hunk 2 above)
Comment (rajaths010494) [reply python-sdk/example_dags/example_load_file.py]: modiifed

File: .github/workflows/ci-python-sdk.yaml [hunk 4]
```diff
@@ -130,6 +133,10 @@ jobs:
           --name postgres
         ports:
           - 5432:5432
+      sftp:
+        image: rajaths1494/sftp-docker
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .github/workflows/ci-python-sdk.yaml:137]: i saw postgres also been the same so i followed the same way. i can check if i have access to push the iage to registry and try to pull from there

File: .github/workflows/ci-python-sdk.yaml [hunk 5]
```diff
@@ -249,6 +256,10 @@ jobs:
           --health-retries 5
         ports:
           - 5432:5432
+      sftp:
+        image: rajaths1494/sftp-docker
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .github/workflows/ci-python-sdk.yaml:260]: i saw postgres also been the same so i followed the same way. i can check if i have access to push the iage to registry and try to pull from there

File: .github/workflows/ci-python-sdk.yaml [hunk 6]
```diff
@@ -320,6 +331,10 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .github/workflows/ci-python-sdk.yaml:335]: i saw postgres also been the same so i followed the same way. i can check if i have access to push the iage to registry and try to pull from there

File: python-sdk/src/astro/files/locations/sftp/sftp.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    @property
+    def transport_params(self) -> dict:
+        """get SFTP credentials for remote file system"""
+        client = self.hook.get_connection(self.conn_id)
+        extra_options = client.extra_dejson
+        if "key_file" in extra_options:
+            key_file = extra_options.get("key_file")
+            return {"connect_kwargs": {"key_filename": key_file}}
+        elif client.password:
+            return {"connect_kwargs": {"password": client.password}}
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply python-sdk/src/astro/files/locations/sftp/sftp.py:30]: yes i will add an exception here

File: .github/workflows/ci-python-sdk.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaxil) [reply .github/workflows/ci-python-sdk.yaml:137]: Good catch about postgres @rajaths010494 , we should move them to https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-docker-registry as they are only used in CI

File: .github/workflows/ci-python-sdk.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaxil) [reply .github/workflows/ci-python-sdk.yaml:260]: https://github.com/astronomer/astro-sdk/pull/1360#discussion_r1053270292

File: python-sdk/src/astro/files/locations/sftp/sftp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rajaths010494) [reply python-sdk/src/astro/files/locations/sftp/sftp.py:31]: added an exception

File: .github/workflows/ci-python-sdk.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .github/workflows/ci-python-sdk.yaml:139]: i have pushed the image to packages in repo 
docker pull [ghcr.io/astronomer/astro-sdk/sftp_docker:latest](http://ghcr.io/astronomer/astro-sdk/sftp_docker:latest)

File: .github/workflows/ci-python-sdk.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .github/workflows/ci-python-sdk.yaml:263]: i have pushed the image to packages in repo
docker pull [ghcr.io/astronomer/astro-sdk/sftp_docker:latest](http://ghcr.io/astronomer/astro-sdk/sftp_docker:latest)

**Comment by rajaths010494** [comment]
https://github.com/astronomer/astro-sdk/pull/1481 PR for the same . Since main got merged here and other changes are reflected in this branch

**Pull request closed without merging**